modules/openy_gc_personal_training/openy_gc_personal_training.module

```diff
@@ -182,12 +162,151 @@ function openy_gc_personal_training_mail($key, &$message, $params) {
+    ],
+    'personal-training-instructor' => [
+      'name' => [
+        'name' => t('Instructor name'),
+        'description' => t('Name of the instructor conducting the personal training.'),
+      ],
+    ],
+    'personal-training-customer' => [
+      'name' => [
+        'name' => t('Customer name'),
+        'description' => t('Name of the customer of the personal training.'),
+      ],
+    ],
+  ];
+
+  return $info;
+}
+
+/**
+ * Implements hook_tokens().
+ */
+function openy_gc_personal_training_tokens($type, $tokens, array $data, array $options, BubbleableMetadata $bubbleable_metadata) {
+  $replacements = [];
+
+  if (
+    $type === 'personal-training'
+    && !empty($data['personal_training'])
+    && $data['personal_training'] instanceof PersonalTrainingInterface
+  ) {
+    foreach ($tokens as $name => $original) {
+      switch ($name) {
+        case 'title':
+          $replacements[$original] = $data['personal_training']->get('title')->value;
+          break;
+
+        case 'start-date':
+          $personal_training = $data['personal_training'];
+          $timezone = date_default_timezone_get();
+
+          $meeting_start_date = '';
+          if ($personal_training->bundle() === 'personal_training') {
+            $startDt = DrupalDateTime::createFromFormat(
+              DateTimeItemInterface::DATETIME_STORAGE_FORMAT,
+              $personal_training->get('date')->value,
+              DateTimeItemInterface::STORAGE_TIMEZONE
+            );
+            $startDt->setTimezone(timezone_open($timezone));
+            $meeting_start_date = $startDt->format('r');
+          }
+          elseif ($personal_training->bundle() === 'training_series') {
+            $dates = \Drupal::service('personal_training_series.manager')->getTrainingsDates($personal_training);
+            $formatted_start_date = $dates['start']->format('r');
+            $formatted_end_date = $dates['end']->format('r');
+            $meeting_start_date = "$formatted_start_date - $formatted_end_date";
+          }
+
+          $replacements[$original] = $meeting_start_date;
+          break;
+      }
+    }
+  }
+
+  if (
+    $type === 'personal-training-instructor'
+    && !empty($data['personal_training_instructor'])
+    && $data['personal_training_instructor'] instanceof UserInterface
+  ) {
+    foreach ($tokens as $name => $original) {
+      switch ($name) {
+        case 'name':
+          $replacements[$original] = $data['personal_training_instructor']->getDisplayName();
+          break;
+      }
+    }
+  }
+
+  if (
+    $type === 'personal-training-customer'
+    && !empty($data['personal_training_customer'])
+    && $data['personal_training_customer'] instanceof UserInterface
+  ) {
+    foreach ($tokens as $name => $original) {
+      switch ($name) {
+        case 'name':
+          $replacements[$original] = $data['personal_training_customer']->getDisplayName();
+          break;
+      }
+    }
+  }
+
+  return $replacements;
```

Comment:
Reading this code and remember when I was a child and was coding for Drupal 5 :)

(https://media.giphy.com/media/lkzQSklzkfqUzSd4j3/giphy.gif)